### PR TITLE
Rewrite Socket.ConnectAsync for DNS with async/await

### PR DESCRIPTION
**Pull request by stephentoub**
Rips out all of the APM code that was previously used to implement this and replaces it with {Value}Task-based async/await implementations.

While there's a small perf benefit to this (maybe a ~10% allocation reduction in a basic socket.ConnectAsync), the main benefit is lots of code getting removed and progress towards removal of IAsyncResult from the implementation entirely.0
image
There's also more that could be cleaned up, and we could streamline the tasks-based implementation to skip the SocketAsyncEventArgs entirely, but I haven't done that here.

cc: @geoffkizer, @MihaZupan

### CODE REVIEW

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs:170]: I assume the point here is to avoid allocation when there are no addresses?

Seems like it would be clearer to just explicitly check for an empty array and then bail early from the method. That way endPoint doesn't have to be nullable.

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs:170]: No,  it's to allocate the end point once with the first address and then reuse that same instance with each subsequent address.

Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs:170]: Yeah, sorry, makes sense.

**Review by geoffkizer — APPROVED** [review]
LGTM. Nice work here. 

@MihaZupan should probably review the telemetry stuff.

**Review by antonfirsov — COMMENTED** [review]
Shouldn't we aim for extending test coverage during these refactors? I think we still may have many cornerish cases uncovered, and we need to avoid issues like #41585 which was a result of an overlook in #40476 + an existing test gap.

Although the product code seems to deal with compatibility, a new variant of `ConnectGetsCanceledByDispose` that tries to connect to a DNS endpoint would nice IMO, to cover the DNS case better.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs:246]: Because there's no need to pay for the objects for all of these addresses on any first use.

**Comment by stephentoub** [comment]
> Shouldn't we aim for extending test coverage during these refactors? I think we still may have many cornerish cases uncovered

What are you proposing?  We're not going to wait to do this until we hit 100% code coverage.  Are there specific areas you're concerned about that this touches that are missing tests?  Are there issues for those?  Thanks.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs:403]: From my perspective it was a bug; ConnectAsync throws NotSupportedException here, and the SocketException that was being thrown is "The system detected an invalid pointer address in attempting to use a pointer argument in a call"... basically the implementation was flawed and giving the OS bad inputs.

We can mark it as a breaking change if you like, of course.  I don't have strong feelings about that.

**Comment by antonfirsov** [comment]
> What are you proposing?

I'm proposing that -- whenever we do a refactor -- we should review and sometimes extend coverage around the specific feature set we touch.

> Are there specific areas you're concerned about that this touches that are missing tests?

I think in this particular case, extending `ConnectGetsCanceledByDispose` to `DnsEndPoint` seems like a reasonable addition.

> Are there issues for those?

No. The problem is that we don't really know where the mines are, but it doesn't mean there are none. My idea is to be proactive instead of being reactive like it happened with #40476/#41585. Being conscious about tests during refactors seems like a cheap and realistic "best effort" way to do so.

Alternatively, we can dedicate resources to review the tests, and open+discuss+triage a bunch of issues around them, but my concern is that this process would be too expensive, won't get the attention it deserves, and in the end of the day we'll end up keeping the debt.

**Comment by MihaZupan** [comment]
The new socket telemetry tests do exercise `ConnectGetsCanceledByDispose` on a `DnsEndPoint` (albeit in outerloop)

**Comment by stephentoub** [comment]
> I'm proposing that -- whenever we do a refactor -- we should review and sometimes extend coverage around the specific feature set we touch.

I agree. Which is why I ran code coverage locally to understand what the gaps are, and at least from those coverage sections, I'm not seeing anything meaningful (the only gap I see there is validating behavior when the first address is unreachable, and I don't think we currently have a great way to test that in CI).  Obviously code coverage only tells part of the story, but I'm not sure how to do better.  Are you suggesting building out some checklist of functionality and reviewing every test in System.Net.Sockets.Tests to determine whether it exercises everything relevant?  Or something else?

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs
Comment (antonfirsov) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs:403]: I'm also unsure. "breaking change" seems like a big hammer, if we add it for all smaller changes, where the chance to break someone is very low, we may end up having a very noisy breaking change list.

It might be reasonable to add a softer, "needs doc update" label to make sure we keep our docs in sync, if smaller changes of this kind are common.

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs:403]: > It might be reasonable to add a softer, "needs doc update" label, if smaller changes of this kind are common.

Seems reasonable to me.  @karelz?

**Comment by antonfirsov** [comment]
> Are you suggesting building out some checklist of functionality and reviewing every test in System.Net.Sockets.Tests to determine whether it exercises everything relevant as part of any PR that does any refactoring?

This sounds like a lot of work. I would do the same, but only as part of individual refactor PR-s/epics, narrowing down the checklist to the feature sets affected by the specific refactor.

> But I'm not sure how to act on your feedback.

I think in this particular PR the only open question is if my suggestion for a separate `ConnectGetsCanceledByDispose(dnsEndpoint)` makes sense. My opinion is that having individual tests for individual bits of functionality is safer, and more maintainable, so relying on the fact that other, telemetry-focused tests already exercise the cancellation path is not enough.

**Comment by stephentoub** [comment]
> a separate ConnectGetsCanceledByDispose(dnsEndpoint)

Added: https://github.com/dotnet/runtime/pull/43661/commits/1b3c4fab066f7e92c6fa713afc2c3a34e85eea45

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs
Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs:403]: It's also worth pointing out that this is an API usage exception -- it indicates a bug in the user's code. We have more flexibility in changing these sorts of exceptions because they shouldn't really ever be handled in a try/catch. It's usually more important to be clear about the root issue here (so the user can fix it) that to maintain backward compat.

That's in contrast to the issue in #41585, which is not an API usage issue.

Comment (carlossanlop) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs:403]: I'll make sure to include the concept of "needs doc update" in our documentation plan.

We are planning to make triple slash comments the source of truth. This is one more argument in favor of doing such change, because documentation would be done directly in the PR modifying the source code behavior.

Right now, since the source of truth is dotnet-api-docs, we should make the change in this file:
https://github.com/dotnet/dotnet-api-docs/blob/master/xml/System.Net.Sockets/Socket.xml

**Comment by geoffkizer** [comment]
Do you want another review here? Not sure what changed since the last push...

**Comment by stephentoub** [comment]
No need.  I rebased to address conflicts with Cory's DNS change and to factor that in in one place.

**Pull request merged**
